src/App.tsx

```diff
@@ -1,18 +1,228 @@
+
+  const areAllCompleted = todosFromTheServer.filter(
+    todoFromServer => !todoFromServer.completed,
+  ).length === 0;
+
+  const filterTodos = (type: Filter) => {
+    setCurrentFilter(type);
+    switch (type) {
+      case Filter.COMPLETED: {
+        const completedTodos = todosFromTheServer.filter(
+          todo => todo.completed,
+        );
+
+        setTodosToShow(completedTodos);
+        break;
+      }
+
+      case Filter.ACTIVE: {
+        const activeTodos = todosFromTheServer.filter(
+          todo => !todo.completed,
+        );
+
+        setTodosToShow(activeTodos);
+        break;
+      }
+
+      default: {
+        setTodosToShow(todosFromTheServer);
+      }
+    }
+  };
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    if (!user) {
+      return;
+    }
+
+    getTodos(user.id)
+      .then(gotTodos => {
+        setTodosFromTheServer(gotTodos);
+        setTodosToShow([...gotTodos]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to get todos');
+        setErrorCount(prev => prev + 1);
+      });
+  }, [user]);
+
+  useEffect(() => {
+    filterTodos(currentFilter);
+  }, [todosFromTheServer]);
+
+  const onFormSubmit = (value: string) => {
+    setIsAdding(true);
+
+    if (value.trim().length === 0) {
+      setErrorMessage("Title can't be empty");
+      setErrorCount(prev => prev + 1);
+      setIsAdding(false);
+
+      return;
+    }
+
+    if (!user) {
+      return;
+    }
+
+    const newTodo = {
+      userId: user.id,
+      title: value,
+      completed: false,
+    };
+
+    setCustomTodo(newTodo);
+
+    addTodo(newTodo).then((todo) => {
+      setTodosFromTheServer(prevTodos => [...prevTodos, todo]);
+    }).catch(() => {
+      setErrorMessage('Unable to add todo');
+      setErrorCount(prev => prev + 1);
+    }).finally(() => {
+      setIsAdding(false);
+      setCustomTodo(null);
+    });
+  };
+
+  const onDeleteTodo = (todoId: number) => {
+    setIsChanging(true);
+    deleteTodo(todoId)
+      .then(() => {
+        setTodosFromTheServer(
+          prevTodos => prevTodos.filter(todo => todo.id !== todoId),
+        );
+      })
+      .catch(() => {
+        setErrorMessage('Unable to delete the todo');
+        setErrorCount(prev => prev + 1);
+      })
+      .finally(() => {
+        setIsChanging(false);
+      });
+  };
+
+  const clearCompleted = () => {
+    setDelitingCompleted(true);
+    const completed = todosFromTheServer.filter(todo => todo.completed);
+
+    Promise.all(completed.map(todo => deleteTodo(todo.id)))
+      .then(() => setTodosFromTheServer(
+        prev => prev.filter(todo => !todo.completed),
+      )).catch(() => {
+        setErrorMessage('Something went wrong deleting the todos');
+        setErrorCount(prev => prev + 1);
+      })
+      .finally(() => setDelitingCompleted(false));
+  };
+
+  const toggleAll = () => {
```

Comment:
loader should be shown in each toggling todo

Reply:
it is shown, have you checked it?